parallel_algorithms/P3732/numeric_ranges_algorithms.md

```diff
@@ -0,0 +1,1256 @@
+`zip_view` and operates on tuples of elements. `zip_transform_view` is a convenient way to express this combination;
+applying `reduce` to `zip_transform_view` gives the necessary result (code examples are shown below).
+
+#### Study `ranges::transform` for design hints
+
+Questions about transforms and projections suggest studying `ranges::transform` for design hints.  This leads us to two
+more questions.
+
+1. If transforms and projections are equivalent, then why does `std::ranges::transform` take an optional projection?
+2. If binary transform is equivalent to unary transform of a `zip_transform_view`, then why does binary
+`std::ranges::transform` exist?
+
+##### Binary transform
+
+It can help to look at examples. The code below shows the same binary transform computation done in two different ways:
+without projections and with projections.
+
+::: cmptable
+
+> `transform` without and with projections
+
+### Without projections
+```c++
+struct foo {};
+std::vector<std::tuple<int, foo, std::string>> v1{
+  {5, {}, "five"}, {7, {}, "seven"}, {11, {}, "eleven"}};
+std::vector<std::pair<int, std::string>> v2{
+  {13, "thirteen"}, {17, "seventeen"}, {19, "nineteen"}};
+std::vector<int> out(std::from_range, std::views::repeat(0, 3));;
+std::vector<int> expected{65, 119, 209};
+
+// Without projections: Big, opaque lambda
+std::ranges::transform(v1, v2, out.begin(),
+  [] (auto x, auto y) { return get<0>(x) * get<0>(y); });
+assert(out == expected);
+```
+
+### With projections
+```cpp
+struct foo {};
+std::vector<std::tuple<int, foo, std::string>> v1{
+  {5, {}, "five"}, {7, {}, "seven"}, {11, {}, "eleven"}};
+std::vector<std::pair<int, std::string>> v2{
+  {13, "thirteen"}, {17, "seventeen"}, {19, "nineteen"}};
+std::vector<int> out(std::from_range, std::views::repeat(0, 3));;
+std::vector<int> expected{65, 119, 209};
+
+// With projections: More readable
+std::ranges::transform(v1, v2, out.begin(),
+  std::multiplies{}, get_element<0>{}, get_element<0>{});
+assert(out2 == expected);
+```
+:::
+
+The code without projections using a single big lambda to express the binary operation. Users have to read the big lambda
+to see what it does. So does the compiler, which can hinder optimization if it's not good at inlining.
+In contrast, the version with projections lets users read out loud what it does.
+It also separates the "selection" or "query" part of the transform from the "arithmetic" or "computation" part. The power of
+the ranges abstraction is that users can factor computation on a range from the logic to iterate over that range. It's
+natural to extend this separation to selection logic as well.
+
+##### Unary transform
+
+It's harder to avoid a lambda, as the function that does an operation, in the unary `transform` case.  Most of the named
+C++ Standard Library arithmetic function objects are binary.  Currying them into unary functions in C++ requires either

```

Comment:
```suggestion
Without projections, it's harder to avoid use of lambdas in the unary `transform` case.  Most of the named
C++ Standard Library arithmetic function objects are binary.  Currying them into unary functions in C++ requires either

```